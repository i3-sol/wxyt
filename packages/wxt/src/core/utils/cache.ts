import fs, { ensureDir } from 'fs-extra';
import { FsCache } from '../../types';
import { dirname, resolve } from 'path';
import { writeFileIfDifferent } from './fs';

/**
 * A basic file system cache stored at `<srcDir>/.wxt/cache/<key>`. Just caches a string in a
 * file for the given key.
 *
 * @param srcDir Absolute path to source directory. See `InternalConfig.srcDir`
 */
export function createFsCache(wxtDir: string): FsCache {
  const getPath = (key: string) =>
    resolve(wxtDir, 'cache', encodeURIComponent(key));

  return {
    async set(key: string, value: string): Promise<void> {
      const path = getPath(key);
      await ensureDir(dirname(path));
      await writeFileIfDifferent(path, value);
    },
    async get(key: string): Promise<string | undefined> {
      const path = getPath(key);
      try {
        return await fs.readFile(path, 'utf-8');
      } catch {
        return undefined;
      }
    },
  };
}
