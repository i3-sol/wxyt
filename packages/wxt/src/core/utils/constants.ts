/**
 * Module ID used to build the background in dev mode if the extension doesn't include a background
 * script/service worker.
 */
export const VIRTUAL_NOOP_BACKGROUND_MODULE_ID = 'virtual:user-background';
