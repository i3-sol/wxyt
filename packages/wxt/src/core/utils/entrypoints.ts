import {
  Entrypoint,
  PerBrowserOption,
  ResolvedPerBrowserOptions,
  TargetBrowser,
} from '../../types';
import path, { relative, resolve } from 'node:path';
import { normalizePath } from './paths';

export function getEntrypointName(
  entrypointsDir: string,
  inputPath: string,
  // type: Entrypoint['type'],
): string {
  const relativePath = path.relative(entrypointsDir, inputPath);
  // Grab the string up to the first . or / or \\
  const name = relativePath.split(/[\.\/\\]/, 2)[0];

  return name;
}

export function getEntrypointOutputFile(
  entrypoint: Entrypoint,
  ext: string,
): string {
  return resolve(entrypoint.outputDir, `${entrypoint.name}${ext}`);
}

/**
 * Return's the entrypoint's output path relative to the output directory. Used for paths in the
 * manifest and rollup's bundle.
 */
export function getEntrypointBundlePath(
  entrypoint: Entrypoint,
  outDir: string,
  ext: string,
): string {
  return normalizePath(
    relative(outDir, getEntrypointOutputFile(entrypoint, ext)),
  );
}

/**
 * Given an entrypoint option, resolve it's value based on a target browser.
 */
export function resolvePerBrowserOption<T>(
  option: PerBrowserOption<T>,
  browser: TargetBrowser,
): T {
  if (typeof option === 'object' && !Array.isArray(option))
    return (option as any)[browser];
  return option;
}

/**
 * Given an entrypoint option, resolve it's value based on a target browser.
 *
 * defaultIcon is special, it's the only key that's a record, which can confuse this function. So
 * it's been manually excluded from resolution.
 */
export function resolvePerBrowserOptions<
  T extends Record<string, any>,
  TKeys extends keyof T,
>(options: T, browser: TargetBrowser): ResolvedPerBrowserOptions<T, TKeys> {
  // @ts-expect-error: Object.entries is untyped.
  return Object.fromEntries(
    Object.entries(options).map(([key, value]) => [
      key,
      key === 'defaultIcon' ? value : resolvePerBrowserOption(value, browser),
    ]),
  );
}

/**
 * Returns true when the entrypoint is an HTML entrypoint.
 *
 * Naively just checking the file extension of the input path.
 */
export function isHtmlEntrypoint(entrypoint: Entrypoint): boolean {
  return entrypoint.inputPath.endsWith('.html');
}
