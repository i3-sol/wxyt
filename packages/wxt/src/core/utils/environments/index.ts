export * from './browser-environment';
export * from './extension-environment';
