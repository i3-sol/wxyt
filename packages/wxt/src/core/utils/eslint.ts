export async function getEslintVersion(): Promise<string[]> {
  try {
    const require = (await import('node:module')).default.createRequire(
      import.meta.url,
    );
    const { ESLint } = require('eslint');
    return ESLint.version?.split('.') ?? [];
  } catch (error) {
    return [];
  }
}
