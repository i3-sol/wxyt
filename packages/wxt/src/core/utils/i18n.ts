export interface Message {
  name: string;
  message: string;
  description?: string;
}

const predefinedMessages = {
  '@@extension_id': {
    message: '<browser.runtime.id>',
    description:
      "The extension or app ID; you might use this string to construct URLs for resources inside the extension. Even unlocalized extensions can use this message.\nNote: You can't use this message in a manifest file.",
  },
  '@@ui_locale': {
    message: '<browser.i18n.getUiLocale()>',
    description: '',
  },
  '@@bidi_dir': {
    message: '<ltr|rtl>',
    description:
      'The text direction for the current locale, either "ltr" for left-to-right languages such as English or "rtl" for right-to-left languages such as Japanese.',
  },
  '@@bidi_reversed_dir': {
    message: '<rtl|ltr>',
    description:
      'If the @@bidi_dir is "ltr", then this is "rtl"; otherwise, it\'s "ltr".',
  },
  '@@bidi_start_edge': {
    message: '<left|right>',
    description:
      'If the @@bidi_dir is "ltr", then this is "left"; otherwise, it\'s "right".',
  },
  '@@bidi_end_edge': {
    message: '<right|left>',
    description:
      'If the @@bidi_dir is "ltr", then this is "right"; otherwise, it\'s "left".',
  },
};

/**
 * Get a list of all messages and their metadata from JSON file contents.
 *
 * @param messagesJson The contents of a `_locales/en/messages.json` file.
 */
export function parseI18nMessages(messagesJson: object): Message[] {
  return Object.entries({
    ...predefinedMessages,
    ...messagesJson,
  }).map<Message>(([name, details]) => ({
    name,
    ...details,
  }));
}
