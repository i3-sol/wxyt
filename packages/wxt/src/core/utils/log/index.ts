export * from './printBuildSummary';
export * from './printFileList';
export * from './printHeader';
export * from './printTable';
