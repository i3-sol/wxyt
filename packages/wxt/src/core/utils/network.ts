import dns from 'node:dns';
import { ResolvedConfig } from '../../types';
import { withTimeout } from './time';

function isOffline(): Promise<boolean> {
  const isOffline = new Promise<boolean>((res) => {
    dns.resolve('google.com', (err) => {
      if (err == null) {
        res(false);
      } else {
        res(true);
      }
    });
  });
  return withTimeout(isOffline, 1e3).catch(() => true);
}

export async function isOnline(): Promise<boolean> {
  const offline = await isOffline();
  return !offline;
}

/**
 * Fetches a URL with a simple GET request. Grabs it from cache if it doesn't exist, or throws an
 * error if it can't be resolved via the network or cache.
 */
export async function fetchCached(
  url: string,
  config: ResolvedConfig,
): Promise<string> {
  let content: string = '';

  if (await isOnline()) {
    const res = await fetch(url);
    if (res.status < 300) {
      content = await res.text();
      await config.fsCache.set(url, content);
    } else {
      config.logger.debug(
        `Failed to download "${url}", falling back to cache...`,
      );
    }
  }

  if (!content) content = (await config.fsCache.get(url)) ?? '';
  if (!content)
    throw Error(
      `Offline and "${url}" has not been cached. Try again when online.`,
    );

  return content;
}
