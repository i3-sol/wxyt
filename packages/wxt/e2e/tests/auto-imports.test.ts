import { describe, it, expect } from 'vitest';
import { TestProject } from '../utils';

describe('Auto Imports', () => {
  describe('imports: { ... }', () => {
    it('should generate a declaration file, imports.d.ts, for auto-imports', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.serializeFile('.wxt/types/imports.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/types/imports.d.ts
          ----------------------------------------
          // Generated by wxt
          export {}
          declare global {
            const ContentScriptContext: typeof import('wxt/client')['ContentScriptContext']
            const InvalidMatchPattern: typeof import('wxt/sandbox')['InvalidMatchPattern']
            const MatchPattern: typeof import('wxt/sandbox')['MatchPattern']
            const browser: typeof import('wxt/browser')['browser']
            const createIframeUi: typeof import('wxt/client')['createIframeUi']
            const createIntegratedUi: typeof import('wxt/client')['createIntegratedUi']
            const createShadowRootUi: typeof import('wxt/client')['createShadowRootUi']
            const defineBackground: typeof import('wxt/sandbox')['defineBackground']
            const defineConfig: typeof import('wxt')['defineConfig']
            const defineContentScript: typeof import('wxt/sandbox')['defineContentScript']
            const defineUnlistedScript: typeof import('wxt/sandbox')['defineUnlistedScript']
            const defineWxtPlugin: typeof import('wxt/sandbox')['defineWxtPlugin']
            const fakeBrowser: typeof import('wxt/testing')['fakeBrowser']
            const storage: typeof import('wxt/storage')['storage']
          }
          "
        `);
    });

    it('should include auto-imports in the project', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.serializeFile('.wxt/wxt.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/wxt.d.ts
          ----------------------------------------
          // Generated by wxt
          /// <reference types="wxt/vite-builder-env" />
          /// <reference types="./types/paths.d.ts" />
          /// <reference types="./types/i18n.d.ts" />
          /// <reference types="./types/globals.d.ts" />
          /// <reference types="./types/imports.d.ts" />
          "
        `);
    });
  });

  describe('imports: false', () => {
    it('should not generate a imports.d.ts file', async () => {
      const project = new TestProject();
      project.setConfigFileConfig({
        imports: false,
      });
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.fileExists('.wxt/types/imports.d.ts')).toBe(false);
    });

    it('should not include imports.d.ts in the type references', async () => {
      const project = new TestProject();
      project.setConfigFileConfig({
        imports: false,
      });
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(
        await project.serializeFile('.wxt/wxt.d.ts'),
      ).toMatchInlineSnapshot(
        `
        ".wxt/wxt.d.ts
        ----------------------------------------
        // Generated by wxt
        /// <reference types="wxt/vite-builder-env" />
        /// <reference types="./types/paths.d.ts" />
        /// <reference types="./types/i18n.d.ts" />
        /// <reference types="./types/globals.d.ts" />
        "
      `,
      );
    });
  });

  describe('eslintrc', () => {
    it('should output the globals list for ESLint to consume', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: true,
          },
        },
      });

      expect(await project.serializeFile('.wxt/eslintrc-auto-import.json'))
        .toMatchInlineSnapshot(`
          ".wxt/eslintrc-auto-import.json
          ----------------------------------------
          {
            "globals": {
              "ContentScriptContext": true,
              "InvalidMatchPattern": true,
              "MatchPattern": true,
              "browser": true,
              "createIframeUi": true,
              "createIntegratedUi": true,
              "createShadowRootUi": true,
              "defineBackground": true,
              "defineConfig": true,
              "defineContentScript": true,
              "defineUnlistedScript": true,
              "defineWxtPlugin": true,
              "fakeBrowser": true,
              "storage": true
            }
          }
          "
        `);
    });

    it('should allow customizing the output', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: true,
            filePath: project.resolvePath('example.json'),
            globalsPropValue: 'readonly',
          },
        },
      });

      expect(await project.serializeFile('example.json'))
        .toMatchInlineSnapshot(`
          "example.json
          ----------------------------------------
          {
            "globals": {
              "ContentScriptContext": "readonly",
              "InvalidMatchPattern": "readonly",
              "MatchPattern": "readonly",
              "browser": "readonly",
              "createIframeUi": "readonly",
              "createIntegratedUi": "readonly",
              "createShadowRootUi": "readonly",
              "defineBackground": "readonly",
              "defineConfig": "readonly",
              "defineContentScript": "readonly",
              "defineUnlistedScript": "readonly",
              "defineWxtPlugin": "readonly",
              "fakeBrowser": "readonly",
              "storage": "readonly"
            }
          }
          "
        `);
    });
  });
});
