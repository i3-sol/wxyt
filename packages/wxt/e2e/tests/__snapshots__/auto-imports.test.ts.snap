// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auto Imports > eslintrc > "enabled: 8" should output a JSON config file compatible with ESlint 8 1`] = `
".wxt/eslintrc-auto-import.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": true,
    "InvalidMatchPattern": true,
    "MatchPattern": true,
    "browser": true,
    "createIframeUi": true,
    "createIntegratedUi": true,
    "createShadowRootUi": true,
    "defineBackground": true,
    "defineConfig": true,
    "defineContentScript": true,
    "defineUnlistedScript": true,
    "defineWxtPlugin": true,
    "fakeBrowser": true,
    "storage": true
  }
}
"
`;

exports[`Auto Imports > eslintrc > "enabled: 9" should output a flat config file compatible with ESlint 9 1`] = `
".wxt/eslint-auto-imports.mjs
----------------------------------------
const globals = {
  "ContentScriptContext": true,
  "InvalidMatchPattern": true,
  "MatchPattern": true,
  "browser": true,
  "createIframeUi": true,
  "createIntegratedUi": true,
  "createShadowRootUi": true,
  "defineBackground": true,
  "defineConfig": true,
  "defineContentScript": true,
  "defineUnlistedScript": true,
  "defineWxtPlugin": true,
  "fakeBrowser": true,
  "storage": true
}

export default {
  languageOptions: {
    globals,
    sourceType: "module",
  },
};
"
`;

exports[`Auto Imports > eslintrc > "enabled: true" should output a JSON config file compatible with ESlint 8 1`] = `
".wxt/eslintrc-auto-import.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": true,
    "InvalidMatchPattern": true,
    "MatchPattern": true,
    "browser": true,
    "createIframeUi": true,
    "createIntegratedUi": true,
    "createShadowRootUi": true,
    "defineBackground": true,
    "defineConfig": true,
    "defineContentScript": true,
    "defineUnlistedScript": true,
    "defineWxtPlugin": true,
    "fakeBrowser": true,
    "storage": true
  }
}
"
`;

exports[`Auto Imports > eslintrc > should allow customizing the output 1`] = `
"example.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": "readonly",
    "InvalidMatchPattern": "readonly",
    "MatchPattern": "readonly",
    "browser": "readonly",
    "createIframeUi": "readonly",
    "createIntegratedUi": "readonly",
    "createShadowRootUi": "readonly",
    "defineBackground": "readonly",
    "defineConfig": "readonly",
    "defineContentScript": "readonly",
    "defineUnlistedScript": "readonly",
    "defineWxtPlugin": "readonly",
    "fakeBrowser": "readonly",
    "storage": "readonly"
  }
}
"
`;
